crates/oxc_codegen/src/sourcemap_builder.rs

```diff
@@ -131,12 +136,29 @@ impl SourcemapBuilder {
 
     #[allow(clippy::cast_possible_truncation)]
     fn search_original_line_and_column(&mut self, position: u32) -> (u32, u32) {
-        let result = self
-            .line_offset_tables
-            .lines
-            .partition_point(|table| table.byte_offset_to_start_of_line <= position);
-        let original_line = if result > 0 { result - 1 } else { 0 };
-        let line = &self.line_offset_tables.lines[original_line];
+        let lines = &self.line_offset_tables.lines;
+        let mut idx = self.last_line_lookup;
+
+        // Ensure we have a valid starting point
+        // If position < lines[idx].byte_offset_to_start_of_line, fallback to a full binary search.
+        if position < lines[idx].byte_offset_to_start_of_line {
+            #[cold]
+            fn fallback_binary_search(lines: &[Line], position: u32) -> usize {
+                let mut i =
```

Comment:
There's an optimization you can make in the fallback branch. You already know that `position < lines[idx].byte_offset_to_start_of_line`. So you could reduce the search to cover only `lines[..=idx]` (`..=idx` not `..idx` because you're searching for the line *after* `position`).

Reply:
Should also modify the tests to make sure they check correctness of both branches. At present they're not testing the fallback branch. Running each test case twice - once in forward order, and once in reverse - should do it.

Reply:
good call out.

After testing this with the transformer, I don't think we need binary search here.

I've been benching this against `checker.ts` ( i copied the main function 2x to make the file a bit bigger), and all transformer options enabled.

```
[crates/oxc_codegen/src/sourcemap_builder.rs:105:9] self.total_search_count = 353513
[crates/oxc_codegen/src/sourcemap_builder.rs:105:9] self.total_search_fallback_count = 16
```

This function was called 353,513 times, and 16 of those times, it had to fall back to binary search.

If we dive into those 16x function calls, the diffs between what the line we previously found and the correct line was: `-10, -23, -4, -5, -10, -4, -5, -5, -4, -5, -3, -3, -8, -5, -4, -6`.

Since lines is a **massive** array (`104790` length), we can probably just linear-ly search here.

However given how infrequently this codepath is called, i doubt it'll make a difference

Reply:
i've added some tests, let me know if they cover/don't cover everything

Reply:
I've pushed a commit to change those new tests. I just pushed a commit because it was faster than explaining what I was thinking. But this is your PR. Feel free to revert or change it.